windows/code/DBStorage.cpp

```diff
@@ -351,174 +339,236 @@ winrt::Windows::Foundation::IAsyncAction DBStorage::RunTasks()
+    auto sql =
+        MakeSQLiteParameterizedStatement("DELETE FROM AsyncLocalStorage WHERE key IN ", argCount);
+    auto statement = StatementPtr{nullptr, &sqlite3_finalize};
+    CHECK_SQL_OK(PrepareStatement(db, sql, &statement));
     for (int i = 0; i < argCount; i++) {
-        if (!BindString(db, m_callback, pStmt, i + 1, m_args[i].AsString())) {
-            return;
-        }
+        CHECK(BindString(statement, i + 1, keys[i]));
     }
-    for (auto stepResult = sqlite3_step(pStmt.get()); stepResult != SQLITE_DONE;
-         stepResult = sqlite3_step(pStmt.get())) {
+    for (;;) {
+        auto stepResult = sqlite3_step(statement.get());
+        if (stepResult == SQLITE_DONE) {
+            break;
+        }
         if (stepResult != SQLITE_ROW) {
-            InvokeError(m_callback, sqlite3_errmsg(db));
-            return;
+            return m_errorHandler.AddError(sqlite3_errmsg(db));
         }
     }
-    std::vector<winrt::JSValue> callbackParams;
-    callbackParams.push_back(winrt::JSValueArray());
-    m_callback(callbackParams);
+    return true;
 }
 
-void DBStorage::DBTask::getAllKeys(sqlite3 *db)
+std::optional<std::vector<std::string>> DBStorage::DBTask::GetAllKeys(sqlite3 *db) noexcept
 {
-    winrt::JSValueArray result;
-    auto getAllKeysCallback = [&](int cCol, char **rgszColText, char **) {
-        if (cCol >= 1) {
-            result.push_back(rgszColText[0]);
+    CHECK(m_errorHandler.GetErrors().empty());
+    std::vector<std::string> result;
+    auto getAllKeysCallback = [&](int columnCount, char **columnTexts, char **) {
+        if (columnCount > 0) {
+            result.emplace_back(columnTexts[0]);
         }
         return SQLITE_OK;
     };
 
-    if (Exec(db, m_callback, u8"SELECT key FROM AsyncLocalStorage", getAllKeysCallback)) {
-        auto writer = winrt::MakeJSValueTreeWriter();
-        result.WriteTo(writer);
-        std::vector<winrt::JSValue> callbackParams;
-        callbackParams.push_back(nullptr);
-        callbackParams.push_back(winrt::TakeJSValue(writer));
-        m_callback(callbackParams);
-    }
+    CHECK(Exec(db, m_errorHandler, "SELECT key FROM AsyncLocalStorage", getAllKeysCallback));
+    return result;
 }
 
-void DBStorage::DBTask::clear(sqlite3 *db)
+std::optional<bool> DBStorage::DBTask::RemoveAll(sqlite3 *db) noexcept
 {
-    if (Exec(db, m_callback, u8"DELETE FROM AsyncLocalStorage")) {
-        std::vector<winrt::JSValue> callbackParams;
-        callbackParams.push_back(nullptr);
-        m_callback(callbackParams);
+    CHECK(m_errorHandler.GetErrors().empty());
+    CHECK(Exec(db, m_errorHandler, "DELETE FROM AsyncLocalStorage"));
+    return true;
+}
+
+// Read KeyValue from a JSON array.
+void ReadValue(const winrt::IJSValueReader &reader,
+               /*out*/ DBStorage::KeyValue &value) noexcept
+{
+    if (reader.ValueType() == winrt::JSValueType::Array) {
+        int index = 0;
+        while (reader.GetNextArrayItem()) {
```

Comment:
For my own education, how does this work? It looks like we have to call `GetNextArrayItem()` at least twice, once for the key and again for the value. Then for some reason, we have to keep calling `GetNextArrayItem()` until it returns `false`? Why is that?

Reply:
You are right, we need to call it only twice for Key and Value. We do it in the cycle in case if data is wrong and someone gave us an array with an unknown number of items. This way the first item will be Key if it is present, the second - Value if it is present, and then we just ignore the rest. Ideally, we would avoid this dance, but I have no way to stop and report an error. This is definitely a flaw in the ReadValue design. I will think how to fix it in future versions of RNW and then we can change it here.